lldb/source/Core/ModuleList.cpp

```diff
@@ -1670,15 +1658,94 @@ static llvm::Expected<bool> loadModuleFromCAS(ConstString module_name,
   loaded.GetArchitecture() = module_spec.GetArchitecture();
   module_spec = loaded;
 
-  LLDB_LOG(GetLog(LLDBLog::Modules), "loading module '{0}' using CASID '{1}'",
-           module_name, cas_id);
+  LLDB_LOG(GetLog(LLDBLog::Modules), "loading module using CASID '{0}'",
+           cas_id);
   return true;
 }
 
+static llvm::Expected<bool> loadModuleFromCAS(llvm::StringRef cas_id,
+                                              const lldb::ModuleSP &nearby,
+                                              ModuleSpec &module_spec) {
+  static llvm::StringMap<bool> g_cache;
+  static std::recursive_mutex g_cache_lock;
+  std::scoped_lock<std::recursive_mutex> lock(g_cache_lock);
+  auto cached = g_cache.find(cas_id);
+  if (cached != g_cache.end())
+    return cached->second;
+  auto result = loadModuleFromCASImpl(cas_id, nearby, module_spec);
+  // Errors are only returned the first time.
+  g_cache.insert({cas_id, result ? *result : false});
+  return result;
+}
+
+llvm::Expected<ModuleList::CAS>
+ModuleList::GetOrCreateCAS(const ModuleSP &module_sp) {
+  if (!module_sp)
+    return llvm::createStringError("no lldb::Module available");
+
+  auto &shared_module_list = GetSharedModuleListInfo();
+  std::scoped_lock<std::mutex> lock(shared_module_list.shared_lock);
+  if (shared_module_list.cas.object_store)
+    return shared_module_list.cas;
+
+  std::set<FileSpec> unique_paths;
+  auto insert_module_path = [&](const ModuleSP &m) -> IterationAction {
+    if (m) {
+      FileSpec path = m->GetFileSpec();
+      path.ClearFilename();
+      unique_paths.insert(path);
+    }
+    return IterationAction::Continue;
+  };
+  insert_module_path(module_sp);
+
+  std::optional<llvm::cas::CASConfiguration> cas_config;
+  for (auto &path : unique_paths)
+    cas_config = GetCASConfiguration(path);
+  if (!cas_config) {
+    unique_paths.clear();
+    insert_module_path(module_sp);
+    if (SymbolFile *sf = module_sp->GetSymbolFile()) {
+      sf->GetDebugInfoModules().ForEach(insert_module_path);
+      for (auto &path : unique_paths)
+        if ((cas_config = GetCASConfiguration(path)))
+          break;
+    }
+  }
+  if (!cas_config &&
+      !ModuleList::GetGlobalModuleListProperties().GetCASOnDiskPath())
```

Comment:
Isn't this a given? Since `GetCASConfiguration()` check this every time (that is kind of redundant for look up above).

Unless you are trying to be thread safe here, but it is also not really thread safe when you have get/set like this.